### PR TITLE
feat: new debugging page

### LINKED ISSUE #4: Document how to debug your contract and dump variables
Add one section about debugging contracts in general

Add the `dump` function to the reference of static available functions

### PR DESCRIPTION
**Pull request by novusnota**
Closes #4

WIP, but almost there.

Preview:
image

### CODE REVIEW

File: pages/book/debug.mdx [hunk 1]
```diff
@@ -1,166 +1,197 @@
...
+
+    If something else happened, it's most likely a logic error in your code, when your expectations didn't match the actual state the contract got in. To resolve that, try stepping through your code and checking state of variables in it via [`dump(){:tact}`][dump] function or alike.
+
+### Examine your assumptions [#approach-2]
+
+Before you investigate a bug or an error, think of the assumptions that made you expect a certain result. Unknown or unclear expectations can get in the way of identifying a problem, even when you're looking right at the cause of the problem.
+
+Here are a few questions to ask yourself to challenge your assumptions:
+
+* Are you using the right API (that is, the right [global static function](/book/functions#global-static-functions) or [extension function](/book/functions#extension-function))? An API that you're using might not do what you think it does, so make sure to consult with the [Reference](/ref) section or ask in the Tact's [Telegram chat][tg].
```
Comment (anton-trunov) [inline pages/book/debug.mdx:31]:
```suggestion
* Are you using the right API (that is, the right [global static function](/book/functions#global-static-functions) or [extension function](/book/functions#extension-function))? An API that you're using might not do what you think it does, so make sure to consult with the [Reference](/ref) section or ask in the Tact [Telegram chat][tg].
```

File: pages/book/debug.mdx [hunk 2]
```diff
@@ -1,166 +1,197 @@
...
-import { sample_Contract } from './output/sample_Contract';
+It can be named anything, but I'll name it `Playground` to convey the right intent. To create it, run the following in the root of your [Blueprint][bp] project and remember to make an empty Tact contract by selecting the proper version from the list:
 
-describe("contract", () => {
-    it("should deploy correctly", async () => {
+```shell
+npx blueprint create Playground
+```
+
+Versions of [Blueprint][bp] starting with $0.20.0$ automatically enable debug mode in `wrappers/` for new contracts, so we only have to adjust the testing suite.
```
Comment (anton-trunov) [inline pages/book/debug.mdx:117]:
```suggestion
Versions of [Blueprint][bp] starting with 0.20.0 automatically enable debug mode in `wrappers/` for new contracts, so we only have to adjust the testing suite.
```

**Comment by anton-trunov** [comment]
What needs to be done here to make "ready for review"?

**Comment by novusnota** [comment]
@anton-trunov those sub-headings & their inner content:

image

I'll be finishing those today, shortly after I finish with Sublime updates. Hopefully, you'll have a bunch of stuff to review tomorrow morning :)

**Comment by anton-trunov** [comment]
In the testing/debugging page we should also cover how to test if a contract's receiver throws an exception. Here is a recent user request:

> Here is a link of my smart contract: https://github.com/kaseen/LuckySixTON/blob/master/contracts/lucky_six.tact
>
> Inside of receive function, i check if input is valid. If it is not valid i throw error with nativeThrow(12345);.
>
> My question is, how can i catch this error when i'm writing unit tests the behavior of function to expect that it should revert. 
>
> For example, in hardhat framework for developing smart contracts in Solidity, there is a function expect(transaction).to.be.revertedWith(errorCode)

**Comment by novusnota** [comment]
> My question is, how can i catch this error when i'm writing unit tests the behavior of function to expect that it should revert.

Sandbox/Blueprint analogue would be to write something like this in tests:

```typescript
expect(res.transactions).toHaveTransaction({
    success: false,
    exitCode: 5
});
```

Where `res` is the result of sending a message to a certain receiver we're testing. I'll add that and more to the debugging page by the end of Tuesday, such that the end page will be comprehensive in many regards.

**Comment by anton-trunov** [comment]
And one more popular request regarding testing: how to simulated time passing

**Comment by novusnota** [comment]
> And one more popular request regarding testing: how to simulate time passing

Hmm, I suppose simple "sleep/wait" will do. And, maybe, adding a separate test group, which won't overwrite the state of the blockchain between the individual tests in that group. On it!

**Comment by anton-trunov** [comment]
Wait, no I meant `blockchain.now`:

```
        ... setting up things
        blockchain = await Blockchain.create();
        ... more setting up things
        blockchain.now = STARTING_TIME;
        ... more tests
        blockchain.now = SOME_LATER_TIME;
        ... more tests
```

**Comment by novusnota** [comment]
Oh, I see. Sure!

**Review by anton-trunov — CHANGES_REQUESTED** [review]
Tremendous improvement!

Let's add `dump` usage example to this page, though. Maybe modify the `play` receiver and show how to observe the results of dumping. I believe we don't have these dump logs anywhere in the Tact docs

**Comment by novusnota** [comment]
Added a usage example to "Common debugging functions", right before the "Enabling debug mode ..."

File: pages/book/debug.mdx [hunk 3]
```diff
@@ -1,166 +1,578 @@
...
+
+In practice, when you write:
+
+```tact {2}
+// ... somewhere appropriate in your contract ...
+dump("To infinity and beyond!");
 ```
 
-## Using `@tact-lang/emulator`
+You'll see the following in the [test logs](#tests) whenever execution gets to that [`dump(){:tact}`][dump] call:
```
Comment (anton-trunov) [inline pages/book/debug.mdx:76]: Let's provide an in-place example here. What are exact steps for me as a Blueprint/Tact user to observe the dumped values in my contract? What do I need to execute to get those? And how exactly the test output would look like? This needs to be as clear as possible, it's one of the most frequently asked questions by the novice users

Comment (novusnota) [reply pages/book/debug.mdx:76]: Got it

Comment (novusnota) [reply pages/book/debug.mdx]: Provided an in-place example and placed reference links to it from earlier in the page

**Review by anton-trunov — APPROVED** [review]
great stuff!

**Comment by anton-trunov** [comment]
Let's publish!

**Pull request merged**